API: Add request metrics for disaster recovery (#13825)
canonical/lxd: doc/reference/provided_metrics.md

  - Number of active warnings
```

(api-rates-metrics)=
## API rates metrics

The API rates metrics include `lxd_api_requests_completed_total` and `lxd_api_requests_ongoing`. These metrics can be consumed by an observability tool deployed externally (for example, the [Canonical Observability Stack](https://charmhub.io/topics/canonical-observability-stack) or another third-party tool) to help identify failures or overload on a LXD server. You can set thresholds on the observability tools for these metrics' values to trigger alarms and take programmatic actions.

> These metrics consider all endpoints in the [LXD REST API](../api), with the exception of the `/` endpoint. Requests using an invalid URL are also counted. Requests against the metrics server are also counted. Both introduced metrics include a label `entity_type` based on the main entity type that the endpoint is operating on.

Reviewer: >  Requests against the metrics server are also counted.

What entity type is used for this?

Author: Since the type is determined by the endpoint, it would be`typeServer`. Same as the regular rest server

  - Number of active warnings
```

(api-rates-metrics)=
## API rates metrics

> The API rates metrics include the `lxd_api_requests_completed_total` and `lxd_api_requests_ongoing`. These metrics can be consumed by an observability tool deployed externally (either Canonical COS or third parties) to help identify failures or overload on a LXD server. Thresholds for these metrics' values can be set on these tools to trigger alarms and take programatic actions.

Reviewer 2: ```suggestion
The API rates metrics include `lxd_api_requests_completed_total` and `lxd_api_requests_ongoing`. These metrics can be consumed by an observability tool deployed externally (for example, the [Canonical Observability Stack](https://charmhub.io/topics/canonical-observability-stack) or another third-party tool) to help identify failures or overload on a LXD server. You can set thresholds on the observability tools for these metrics' values to trigger alarms and take programmatic actions.
```

> These metrics consider all endpoints in the [LXD REST API](api.md), plus the internal endpoints. They also include a label `entity_type` based on the main entity type that the endpoint is operating on. This label can have one of the following values:

Reviewer 2: ```suggestion
These metrics consider all endpoints in the [LXD REST API](api), plus the internal endpoints. They also include a label `entity_type` based on the main entity type that the endpoint is operating on. This label can have one of the following values:
```

> - `instance`

Reviewer 2: ```suggestion

- `instance`
```

See the documentation checks - lists must be surrounded by blank lines.

> - `image`, for endpoints that operate on images or image aliases.
> - `storage_pool`, for endpoints that operate on storage pools, storage volumes, storage backups or storage buckets.
> - `network`, for endpoits that operate on networks, network zones or network ACLs.

Reviewer 2: ```suggestion
- `image`, for endpoints that operate on images or image aliases
- `storage_pool`, for endpoints that operate on storage pools, storage volumes, storage backups or storage buckets
- `network`, for endpoints that operate on networks, network zones or network ACLs
```

Reviewer 2: The same change below - we should be consistent with punctuation within a list. So either all items get a full stop at the end, or none of them do. (And in this case none makes way more sense. :) )

- `operation`
- `project`
- `warning`
- `profile`

> - `identity`, for endpoints that related to authorization.

Reviewer 2: ```suggestion
- `identity`, for endpoints related to authorization
```

- `node`, for endpoints related to clustering.
- `server`, for internal endpoints and any remaining endpoints that don't operate on any of the entity types above.

> `lxd_api_requests_ongoing` consists of the number of requests that are not yet completed by the time the metrics are queried. A request is considered completed when the response is returned to the client and any asynchronous operations spawned by that request are done. `lxd_api_requests_completed_total` consists of the number of completed requests. This metric includes an additional label named `result` based on the outcome of the request. The label can assume on of the following values:

Reviewer 2: ```suggestion
`lxd_api_requests_ongoing` contains the number of requests that are not yet completed by the time the metrics are queried. A request is considered completed when the response is returned to the client and any asynchronous operations spawned by that request are done. `lxd_api_requests_completed_total` contains the number of completed requests. This metric includes an additional label named `result` based on the outcome of the request. The label can have one of the following values:
```

> - `error_server`, for errors on the server side, this includes responses with HTTP status codes from 500 to 599. Any failed asynchronous operations also fall into this category.

Reviewer 2: ```suggestion
- `error_server`, for errors on the server side, including responses with HTTP status codes from 500 to 599. Any failed asynchronous operations also fall into this category.
```

Reviewer 2: For this list, having a full stop at the end of all items makes sense since you need it for this item.

Author: Got it, thanks for the insight!

  - Number of active warnings
```

(api-rates-metrics)=
## API rates metrics

The API rates metrics include `lxd_api_requests_completed_total` and `lxd_api_requests_ongoing`. These metrics can be consumed by an observability tool deployed externally (for example, the [Canonical Observability Stack](https://charmhub.io/topics/canonical-observability-stack) or another third-party tool) to help identify failures or overload on a LXD server. You can set thresholds on the observability tools for these metrics' values to trigger alarms and take programmatic actions.
These metrics consider all endpoints in the [LXD REST API](../api), plus the internal endpoints and the {abbr}`ACME (Automatic Certificate Management Environment)` challenge endpoint. See {ref}`authentication-server-certificate` for more information on ACME. They also include a label `entity_type` based on the main entity type that the endpoint is operating on. This label can have one of the following values:

> - `instance`

Reviewer: @hamistao @markylaing is there somewhere in the docs we can link to for entity types, rather than having to duplicate this list? It feels like it will go out of date easily.

Also, @markylaing is the `node` entity type already exposed to end users? I thought we weren't exposing the "node" concept to end users, and using "member" or "cluster member" instead? e.g. `type ClusterGroup struct`

This feels like something we need to change otherwise we're mixing naming for the same entity type.

Author: I don't think there is such a list, and if there is there would still be the need to clarify that some entity types are collapsed in broader categories here. I suppose just removing this list may be an option here.
